coderd/keyrotate/rotate_internal_test.go

```diff
@@ -0,0 +1,526 @@
+		require.Len(t, kbf[database.CryptoKeyFeatureOidcConvert], 1)
+		// Workspace apps should have been rotated.
+		require.Len(t, kbf[database.CryptoKeyFeatureWorkspaceApps], 2)
+		// No existing key for tailnet resume should've
+		// caused a key to be inserted.
+		require.Len(t, kbf[database.CryptoKeyFeatureTailnetResume], 1)
+
+		oidcKey := kbf[database.CryptoKeyFeatureOidcConvert][0]
+		tailnetKey := kbf[database.CryptoKeyFeatureTailnetResume][0]
+		requireKey(t, oidcKey, database.CryptoKeyFeatureOidcConvert, now, time.Time{}, validKey.Sequence)
+		requireKey(t, tailnetKey, database.CryptoKeyFeatureTailnetResume, now, time.Time{}, deletedKey.Sequence+1)
+
+		newKey := kbf[database.CryptoKeyFeatureWorkspaceApps][0]
+		oldKey := kbf[database.CryptoKeyFeatureWorkspaceApps][1]
+		if newKey.Sequence == rotatedKey.Sequence {
+			oldKey, newKey = newKey, oldKey
+		}
+		requireKey(t, oldKey, database.CryptoKeyFeatureWorkspaceApps, rotatedKey.StartsAt.UTC(), rotatedKey.ExpiresAt(keyDuration).Add(WorkspaceAppsTokenDuration+time.Hour), rotatedKey.Sequence)
+		requireKey(t, newKey, database.CryptoKeyFeatureWorkspaceApps, rotatedKey.ExpiresAt(keyDuration), time.Time{}, rotatedKey.Sequence+1)
+	})
+
+	t.Run("UnknownFeature", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 7
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features:    []database.CryptoKeyFeature{database.CryptoKeyFeature("unknown")},
+		}
+
+		err := kr.rotateKeys(ctx)
+		require.Error(t, err)
+	})
+
+	t.Run("MinStartsAt", func(t *testing.T) {
+		t.Parallel()
+
+		var (
+			db, _       = dbtestutil.NewDB(t)
+			clock       = quartz.NewMock(t)
+			keyDuration = time.Hour * 24 * 5
+			logger      = slogtest.Make(t, nil).Leveled(slog.LevelDebug)
+			ctx         = testutil.Context(t, testutil.WaitShort)
+		)
+
+		now := dbnow(clock)
+
+		kr := &rotator{
+			db:          db,
+			keyDuration: keyDuration,
+			clock:       clock,
+			logger:      logger,
+			features:    []database.CryptoKeyFeature{database.CryptoKeyFeatureWorkspaceApps},
+		}
+
+		expiringKey := dbgen.CryptoKey(t, db, database.CryptoKey{
+			Feature:  database.CryptoKeyFeatureWorkspaceApps,
+			StartsAt: now.Add(-keyDuration),
+			Sequence: 345,
+		})
+
+		err := kr.rotateKeys(ctx)
+		require.NoError(t, err)
+
+		keys, err := db.GetCryptoKeys(ctx)
+		require.NoError(t, err)
+		require.Len(t, keys, 2)
+
+		rotatedKey, err := db.GetCryptoKeyByFeatureAndSequence(ctx, database.GetCryptoKeyByFeatureAndSequenceParams{
+			Feature:  expiringKey.Feature,
+			Sequence: expiringKey.Sequence + 1,
+		})
+		require.NoError(t, err)
+		require.Equal(t, now.Add(defaultRotationInterval*3), rotatedKey.StartsAt.UTC())
+	})
```

Comment:
One more slightly insidious edge case:

 2 Workspace apps keys in the database

seq 19, expiresAt 2 hours ago, valid secret
seq 20, deletesAt now, NULL secret (e.g. manually deleted)

We should insert a new key with sequence 21, startsAt 30 minutes from now.
We should set key 19 deletesAt to 1h5m from _now_, since as far as we know, it could be being used to mint tokens right now.

Reply:
I fixed it so that `deletes_at` is dependent on the `starts_at` of the new key. So the `deletes_at` would be `1hr35m` from now.